src/librustc_mir/transform/const_prop.rs

```diff
@@ -521,31 +521,37 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
         None
     }
 
+    /// Check for overflow for a unary op. Like check_binary_op, safe to call even if the operand
+    /// needs substs. In that case, no work is done.
     fn check_unary_op(
         &mut self,
         op: UnOp,
         arg: &Operand<'tcx>,
         source_info: SourceInfo,
     ) -> Option<()> {
-        if self.use_ecx(|this| {
-            let val = this.ecx.read_immediate(this.ecx.eval_operand(arg, None)?)?;
-            let (_res, overflow, _ty) = this.ecx.overflowing_unary_op(op, val)?;
-            Ok(overflow)
-        })? {
-            // `AssertKind` only has an `OverflowNeg` variant, so make sure that is
-            // appropriate to use.
-            assert_eq!(op, UnOp::Neg, "Neg is the only UnOp that can overflow");
-            self.report_assert_as_lint(
-                lint::builtin::ARITHMETIC_OVERFLOW,
-                source_info,
-                "this arithmetic operation will overflow",
-                AssertKind::OverflowNeg,
-            )?;
+        if !arg.needs_subst() {
```

Comment:
Please avoid rightwards-drift, and instead do something like
```rust
if needs_subst { return Some(()); }
```

Reply:
(Also applies to `check_binary_op`.)

Reply:
Don't all of these `needs_subst` changes become obsolete with https://github.com/rust-lang/rust/pull/71386#discussion_r412942498?

Reply:
Yes.

I wrote these comments before you showed up in this PR.